### ISSUE TITLE
Stream completes from incorrectly after caught error

### ISSUE THREAD

**Issue by danielrasmuson** [ReactiveX/rxjs#2574]
**RxJS version:**
5.3.0

**Code to reproduce:**
```javascript
// Will run forever
Rx.Observable.interval(1000)
  .flatMap(() =>new Rx.Observable.create((obs) => obs.error('hi'))
    .catch((err) => Rx.Observable.of(err)))
  .subscribe(() => console.log('next forever'));

// Will emit once and complete
Rx.Observable.interval(1000)
  .flatMap(() =>new Rx.Observable.create((obs) => obs.error('hi')))
  .catch((err) => Rx.Observable.of(err))
  .subscribe(() => console.log('next once'));
```

**Expected behavior:**
Since I'm catching the error they should both emit forever.

**Additional information:**
Here is my scenario in the real world. I have a function..

```javascript
runBatFile('record_screen.bat')
   .subscribe(
     (progressData) => {}, // fps and size information
     (errorData) => {}, // frame was missed, memory is getting high
     () => {} // recording has stopped 
   )
```

And because of the stated `.catch` problem above I have no easy method of continuing the video capture if a frame was missed. The first time we have a caught error `runBatFile().catch()` it will complete the stream and stop listening. Therefore, I'm forced to modify the `runBatFile` function to `.next` all results.

**Comment by kwonoj** [comment]
It is important `when to catch`. Behavior of `catch` is supposed to intecept `onError` handler, and replace original observable into supplied observable.(https://github.com/ReactiveX/rxjs/blob/e38711379f357f55290ead7fabcf8c70c5b31961/spec/operators/catch-spec.ts#L17-L24) Means, Observable still follows error semantics when error is thrown - Observable itself is being terminated as soon as error occurred, then catch only changes handling of those errorneous state.

So in first snippet, source observable `Observable.interval(1000)` continues, as inner observable catches its own error and higher order observable never get error, point of view from `interval` there isn't any error occurred. In opposite, below snippet bubbles up error into higher order, makes observable terminates by handling it.

In short, you may not able to expect `catch` will continue original stream as-is. If you'd like to have continuation of streams, you may either need to look `retry`, or `onErrorResumeNext` per each usecase, or sometimes you can consider `materialize` inner stream to prevent any error being thrown as materialize will emit metadata of observable values instead of actually throwing it out, so you can filter / map it as needed.

**Comment by danielrasmuson** [comment]
That's very helpful. Thank you.
